File: DataStructure/70-Climbing-Stairs.cpp
#include <unordered_map>
class Solution {
public:
    unordered_map<int, int> dp;
    int climbStairs(int n) {
        if (n <= 0) return 1;
        if (n == 1) return 1;

        if (dp.count(n) > 0)
        {
            return dp[n];
        }
        dp[n] = climbStairs(n - 1) + climbStairs(n - 2);
        return dp[n];

    }
};

File: DataStructure/70-Climbing-Stairs.py
class Solution(object):
    def climbStairs(self, n, dp = {}):
        \\\
        :type n: int
        :rtype: int
        \\\
        if n <= 0: return 1
        if n == 1: return 1
        if n in dp:
            return dp[n]
        dp[n] = self.climbStairs(n - 1) + self.climbStairs(n-2)
        return dp[n]
        